# Chapter I - Pandas

In this notebook we will cover how to:
- work with the two main data types in `pandas`: `DataFrame` and `Series`
- work with data types in `pandas`, especially strings and dates
- load data from JSON and CSV into a `DataFrame`
- manipulate the columns of a `DataFrame`
- access data in a `DataFrame` by means of indexes and slicing

In [7]:
import pandas as pd
import numpy as np
from random import random

## Motivation

In [6]:
df = pd.read_csv('2024-08-13_data.txt', header=None)
df.columns=['ID', 'Name', 'Math', 'Science', 'English']
df['average'] = df[['Math', 'Science', 'English']].mean(axis=1)
df

ID           Name  Math  Science  English    average
0  101       John Doe    78       85       92  85.000000
1  102     Jane Smith    88       79       85  84.000000
2  103    Emily Davis    91       89       94  91.333333
3  104  Michael Brown    70       75       80  75.000000
4  105  Jessica White    85       93       89  89.000000

## Section (1): Creating a first dataframe from scratch

This section will take you through the steps needed to create a `pandas` DataFrame from scratch.

To create a DataFrame from scratch, you need the values for at least two columns.
Those values are stored in a data type called a `Series`. They can be thought of as the `pandas` version of lists.

A pandas `Series` can be created as follows:

### A) Pandas Series

In [15]:
s = pd.Series([1,2,3])

print(s)
print(' > The type of s is:', type(s))

0    1
1    2
2    3
dtype: int64
 > The type of s is: <class 'pandas.core.series.Series'>


✏️ [Ex.1] 
- ✏️ Create a series called `s` containing 100 random numbers ranging between 0 and 1. You may use the `random` function.
- ✏️ Print the first value of the `Series` you just created.

In [24]:
# step 1: create a list of 100 random values
# step 2: convert that list to a Pandas Series
# step 3: get the first element of that list

mylist = []

for i in range(100):
    mylist.append(random())

s = pd.Series(mylist)
print(s[0])

0.5187530793651812


Each observation in the series has an **index** as well as a set of **values**: they can be accessed via the omonymous properties.
- The data type of the **index** is a `pandas RangeIndex`, akin to a Python `range`.
- The data type of the **values** is a `numpy array`.

✏️ [Ex.2] 
- ✏️ Using the series **index**, print the length of the `Series`
- ✏️ Print the first three elements of the **values** of series `s`.

In [31]:
print(s.values[:3])

[0.51875308 0.23781756 0.33329084]


Pandas `Series` have got useful properties that you can call to easily access information on the data in the Series.
Some of them include:
- `head(n)` and `tail(n)` to access the beginning and end of the series — where `n` is the number of values to get.
- `value_counts()` to show the occurrences of all values in the series. Calling this property returns a `Counter` object, itself contains an `.index` and some `.values` which you can call to access the occurrences' count.
- `min()`, `max()`, `mean()`, `median()` give some basic statistics on the series' data.

✏️ [Ex.3] 
- ✏️ Calculate the range of values in `s`
- ✏️ Find if there are some duplicate values in `s`
- ✏️ Calculate the mean of the first 50 values in `s`

In [40]:
# your solution
s.tail(30).mean()

0.5204953113548016

- Some of you might want to manipulate time data in the form of dates. Pandas is very convenient for the manipulation of dates. 

To do that, you should use pandas appropriate date type, called `Timestamp`.

For example, VE-day can be encoded as such:

In [43]:
print(pd.Timestamp(1945, 5, 8, 20, 10, 56))

1945-05-08 20:10:56


A date can also be encoded as a string, and pandas will do its best to convert it to a timestamp.

Note that it flexibly supports both 'YYYYMMDD' and 'YYYMMDDHHMMSS' 

In [48]:
print(pd.Timestamp('19450508'))
print(pd.Timestamp('19690711025615'))

# What happens if you try to create a Timestamp with a date that doesn't exist? Try it out.

1945-05-08 00:00:00
1969-07-11 02:56:15


The difference between two `Timestamps` is a `Timedelta` object. The number of days contained in the time difference can be accessed through the eponymous property:

In [59]:
print((pd.Timestamp('19690711025615') - pd.Timestamp('-19450508')).days)

1429623


A date can be shifted simply by adding to it a `Timedelta`:

In [60]:
print(pd.Timestamp('19450508')+pd.Timedelta('55 days 2 hours 15 minutes 10 seconds'))

1945-07-02 02:15:10


✏️ [Ex.4] 
- ✏️ Create a list of pandas `Timestamps` of all the days between the 24th May 1819 and the 22nd January 1901.
- ✏️ By converting this list into a pandas `Series`, get the median day of this time interval.


In [75]:
start = pd.Timestamp('18190524')
end = pd.Timestamp('19010122')

number_of_days = (end - start).days

all_days = [start]

for i in range(number_of_days):
    new_day = all_days[-1] + pd.Timedelta('1 day')
    all_days.append(new_day)

all_days_series = pd.Series(all_days)
all_days_series.mean()

pd.date_range(start=start, end=end, freq="D")

DatetimeIndex(['1819-05-24', '1819-05-25', '1819-05-26', '1819-05-27',
               '1819-05-28', '1819-05-29', '1819-05-30', '1819-05-31',
               '1819-06-01', '1819-06-02',
               ...
               '1901-01-13', '1901-01-14', '1901-01-15', '1901-01-16',
               '1901-01-17', '1901-01-18', '1901-01-19', '1901-01-20',
               '1901-01-21', '1901-01-22'],
              dtype='datetime64[ns]', length=29829, freq='D')

### B) Pandas DataFrames

What is a `pandas.DataFrame`? Think of it as an in-memory spreadsheet that you can analyse and manipulate programmatically.

A `DataFrame` is a collection of `Series` having the same length and whose indexes are in sync. A *collection* means that each column of a dataframe is a series

Let's create a toy `DataFrame` by hand. 

In [88]:
dates = [pd.Timestamp(1970, 5, 23), pd.Timestamp(1978, 7, 14), pd.Timestamp(1986, 3, 14), pd.Timestamp(1993, 1, 1), pd.Timestamp(1998, 7, 14)]
events = ['birth', 'anniversary', 'wedding', 'wedding', 'anniversary']


From those two lists, you can create a `DataFrame` by passing to `pd.DataFrame` a dictionary:

In [90]:
toy_df = pd.DataFrame({
    "date": dates,
    "event": events
})

# What do you expect when dates and events are changed from lists to Series? Try it out.
# What will happen if the lists are of different lengths?

In [92]:
toy_df

date        event
0 1970-05-23        birth
1 1978-07-14  anniversary
2 1986-03-14      wedding
3 1993-01-01      wedding
4 1998-07-14  anniversary

You can check that the `DataFrame` has been properly constructed. Notice how it is indeed of a tabular shape. To extract its length, you can use `len(DataFrame)`. 

In [97]:
print('> This DataFrame has length:', len(toy_df))
display(toy_df)

> This DataFrame has length: 5


date        event
0 1970-05-23        birth
1 1978-07-14  anniversary
2 1986-03-14      wedding
3 1993-01-01      wedding
4 1998-07-14  anniversary

✏️ [Ex.5] 
- ✏️ Create a list of pandas `Timestamps` of 200 random dates between the 1900 and 2000.
- ✏️ Create a list of 200 events taken *at random* among ['birth', 'anniversary', 'wedding']. The numpy function `np.random.choice()` might help.
- ✏️ Combine those two lists to create a dataframe
- ✏️ Get the count of occurrences of all **events** in the dataframe


In [129]:
start = pd.Timestamp('19000101')
end = pd.Timestamp('20000101')

number_of_days = (end - start).days

timedeltas = [randrange(0, number_of_days) for i in range(200)]
random_days = [start + pd.Timedelta(f'{days} day') for days in timedeltas]


random_events = np.random.choice(['birth', 'anniversary', 'wedding'], 200)

df = pd.DataFrame({'date': random_days, 'event':random_events})

In [135]:
df

date        event
0   1972-11-16  anniversary
1   1936-05-28      wedding
2   1917-10-23        birth
3   1935-04-26  anniversary
4   1945-06-01  anniversary
..         ...          ...
195 1921-10-19  anniversary
196 1929-03-08  anniversary
197 1956-05-26  anniversary
198 1971-04-24  anniversary
199 1999-03-10        birth

[200 rows x 2 columns]

Once the `DataFrame` exists, you can add a column in exactly the same way you would define a new key/value pair in a dictionnary:

`dataframe_name['new_column'] = values`


Here, the datatype of `values` is quite flexible as `Pandas` allows many inputs: `pandas.Series` as we've seen before, but also `numpy.array` or even simple `lists`.

The only condition is that the length of the new column should be of the same length as the `DataFrame`.

There is one exception to that rule: if all rows of the new column have the same value, you can just pass that value as input.

✏️ [Ex.6] 
- ✏️ Add a new column named `author_firstname` to the dataframe created in [Ex.5]. 
- ✏️ This new column be input using a list-like variable, containing your first name as many times as there are rows.
- ✏️ Add a new column named `author_lastname` to the dataframe, this time containing your last name, and without using a list-like input.



In [186]:
# your solution

## Section (2): First manipulations of the dataframe

### A) General information on the dataframe

Some first pieces of information on a dataframe are given by the following useful functions: `df.head()`, `df.tail()`, `df.info()`.

The method `info()` gives you information about a dataframe:
- how much space does it take in memory?
- what is the datatype of each column?
- how many records are there?
- how many `null` values does each column contain (!)?


In [138]:
len(toy_df)

5

Alternatively, if you need to know only the number of columns and rows you can use the `.shape` property. 

It is a property, not a method — therefore it should be called without brackets.

Calling the property returns a tuple with 1) number of rows, 2) number of columns.



In [139]:
toy_df.shape

(5, 2)

`head()` prints by first five rows of a dataframe:


In [141]:
toy_df.head(3)


date        event
0 1970-05-23        birth
1 1978-07-14  anniversary
2 1986-03-14      wedding

But the number of lines displayed is a parameter that can be changed:


In [178]:
toy_df.head(2)


date        event
0 1970-05-23        birth
1 1978-07-14  anniversary

`tail()` does the opposite, i.e. prints the last n rows in the dataframe:

In [142]:
toy_df.tail(2)

date        event
3 1993-01-01      wedding
4 1998-07-14  anniversary

You may sometimes want to sort the dataframe based on the values in one column.

To do this, you may use the `.sort_values(<column>)` method. 

The column will then be sorted depending on the datatype:
- numerically (float, integers);
- chronologically (datetimes);
- alphabetically (strings).

In [143]:
toy_df.sort_values('date')

date        event
0 1970-05-23        birth
1 1978-07-14  anniversary
2 1986-03-14      wedding
3 1993-01-01      wedding
4 1998-07-14  anniversary

If you want to invert the sorting (z-to-a, 9-to-1, etc.), use the `ascending=False` argument.

In [146]:
toy_df.sort_values('event', ascending=False)

date        event
2 1986-03-14      wedding
3 1993-01-01      wedding
0 1970-05-23        birth
1 1978-07-14  anniversary
4 1998-07-14  anniversary

### B) Columns and datatype

The columns of a `pandas.DataFrame` can be accessed as follows:

In [148]:
toy_df

date        event
0 1970-05-23        birth
1 1978-07-14  anniversary
2 1986-03-14      wedding
3 1993-01-01      wedding
4 1998-07-14  anniversary

In [152]:
toy_df['date']

0   1970-05-23
1   1978-07-14
2   1986-03-14
3   1993-01-01
4   1998-07-14
Name: date, dtype: datetime64[ns]

 It returns a `pandas.Series`, the type we've seen in the introductory section of this notebook. To access its **values** the property keyword is used:

In [153]:
print(type(toy_df['date']))
print(toy_df['date'].values)

<class 'pandas.core.series.Series'>
['1970-05-23T00:00:00.000000000' '1978-07-14T00:00:00.000000000'
 '1986-03-14T00:00:00.000000000' '1993-01-01T00:00:00.000000000'
 '1998-07-14T00:00:00.000000000']


Each column in a `pandas.DataFrame` has a data type. Being sure that the right datatype is used is essential.

Depending on the nature of the data, its type can be changed using the method `.astype()`. 

For example, changing from a `pandas.Timestamp` to a `str` is possible:

In [157]:
type(toy_df['date'].astype(str)[0])

str

But changing from a `pandas.Timestamp` to a `float` is not possible:

In [162]:
## What do you expect when you run the following?
#print(toy_df['date'].astype(float))

### C) Accessor properties

For certain data types (string, datetime), `pandas` provides a number of common methods that can be called on any series containing values of that type. These methods become available as methods of the series itself within a property — called *accessor* — named after the data type:

- the `.dt.*` accessor contains methods to operate on `datetime` series
- the `str.` accessor contains methods to operate on `str` (string) series.

Accessors are amongst the most convenient features of data manipulation in `pandas`.

They act on a `pandas.Series`, typically the column of a `DataFrame` and return a `pandas.Series` of the same length. The output new series is the result of the element-wise operation on the input series.

Let's start with temporal manipulation:

#### `datetime` accessor

To work with datetime series `pandas` provide a bunch of useful methods to operate on a series: they can be called from the `.dt` property of a datetime series.

They can be used to:
- convert from one timezone to another
- get the day/day name/month/year information from each date
- and much more (see the [documentation]())



In [174]:
toy_df['new_column'] = the_values

toy_df['date'].dt.day_name()

0    Saturday
1      Friday
2      Friday
3      Friday
4     Tuesday
Name: date, dtype: object

✏️ [Ex.7] 
To see this in action: 
- ✏️ Access your dataframe's `date` column
- ✏️ Print for each the corresponding day of the week. To do this, you should use the `datetime` accessor, and use the method `day_name`. This will return a `pandas.Series`.
- ✏️ Add the day of the week as a new column.
- ✏️ Try to do this again in a one-liner.

In [183]:
dates = toy_df['date']
name_days = dates.dt.day_name()
toy_df['day_name'] = name_days

#toy_df['day_name2'] = toy_df['date'].dt.day_name()
toy_df

date        event  day_name day_name2
0 1970-05-23        birth  Saturday  Saturday
1 1978-07-14  anniversary    Friday    Friday
2 1986-03-14      wedding    Friday    Friday
3 1993-01-01      wedding    Friday    Friday
4 1998-07-14  anniversary   Tuesday   Tuesday

#### `str` accessor

Much like the `datetime` accessor, the `str` one is the entry door to many very useful methods that you may need to tidy, process, or analyse your data.

Among other things, you can easily:
- test if the string starts with another string, 
- convert between lower and upper case, 
- determine if the string matches a regular expression,
- replace one substring with one another.

For example, if you want to check if the first three letters of the `event` column are those of "wedding", you can use:

In [193]:
toy_df

date        event  day_name
0 1970-05-23        birth  Saturday
1 1978-07-14  anniversary    Friday
2 1986-03-14      wedding    Friday
3 1993-01-01      wedding    Friday
4 1998-07-14  anniversary   Tuesday

In [195]:
toy_df['day_name'].str.endswith('day')

0    True
1    True
2    True
3    True
4    True
Name: day_name, dtype: bool

You can also chain the accessors. However, remember than the output of an accessor method is a `pandas.Series`. You will therefore need to access `str` again!

For example, if you want first to capitalise a column, before checking whether it starts with the first letters of "wedding", you can do:

In [197]:
toy_df

date        event  day_name
0 1970-05-23        birth  Saturday
1 1978-07-14  anniversary    Friday
2 1986-03-14      wedding    Friday
3 1993-01-01      wedding    Friday
4 1998-07-14  anniversary   Tuesday

In [199]:
#This time, we match with "Wed"

toy_df['event'].str.capitalize().str.startswith('wed')

0    False
1    False
2     True
3     True
4    False
Name: event, dtype: bool

In [210]:
toy_df = toy_df.rename(columns={'day_name':'day'})
toy_df

date        event       day  is_weekend  is_weekday
0 1970-05-23        birth  Saturday        True       False
1 1978-07-14  anniversary    Friday       False        True
2 1986-03-14      wedding    Friday       False        True
3 1993-01-01      wedding    Friday       False        True
4 1998-07-14  anniversary   Tuesday       False        True

✏️ [Ex.8] 
- ✏️ Using the `str` accessors, create a new column `is_weekend` that states if the date fell during a weekend.
- ✏️ The new column should be a boolean (True/False).


In [207]:
# your solution

toy_df['is_weekend'] = toy_df['day_name'].str.startswith('S')

toy_df['is_weekday'] = [not i for i in toy_df['is_weekend'].values]
toy_df

date        event  day_name  is_weekend  is_weekday
0 1970-05-23        birth  Saturday        True       False
1 1978-07-14  anniversary    Friday       False        True
2 1986-03-14      wedding    Friday       False        True
3 1993-01-01      wedding    Friday       False        True
4 1998-07-14  anniversary   Tuesday       False        True

## Section (3): Input/Output

Up until now, we have created and manipulated data from scratch.

However most often, they are created by loading existing data into a dataframe by means of `pandas`' input/output methods:

- Either by loading a complete dataframe, for example if you want to manipulate a CSV file;
- Or by loading the data columns independantly, and combining them into one dataframe.

#### From JSON

A very common data format is JSON, explicit, efficient, and widely used over the internet.

We will take here the example of some data on books from the British Library. We extracted it from the BL as a JSON file. You may face such a scenario in your research.

Loading data from a JSON file is very similar to creating a `DataFrame` from a `dict`, like we've done in Section (1).

This is how one would do it in pure Python:

In [211]:
import json
json_file_path = '../data/bl_books/sample/book_data_sample.json'

# JSON data gets read into a dictionary

with open(json_file_path, 'r') as jsonfile:
    json_data = json.load(jsonfile)
    
books_df = pd.DataFrame(json_data)

In [212]:
books_df.head()

datefield                                         shelfmarks  \
0                1841               [British Library HMNTS 11601.ddd.2.]   
1                1888                [British Library HMNTS 9025.cc.14.]   
2  1847, 48 [1846-48]  [British Library HMNTS C.131.d.16., British Li...   
3                1892               [British Library HMNTS 10351.cc.60.]   
4                1863                 [British Library HMNTS 9078.c.10.]   

               publisher                                              title  \
0      Privately printed  [The Poetical Aviary, with a bird's-eye view o...   
1             Rivingtons  [A History of Greece. Part I. From the earlies...   
2           Punch Office  [The Comic History of England ... With ... col...   
3  Eden, Remington & Co.  [The Cruise of “The Tomahawk”: the story of a ...   
4     Virtue Bros. & Co.        [Scenes from the Drama of European History]   

  edition                          flickr_url_to_book_images     place  \
0          http://www.flickr.com/photos/britishlibrary/ta...  Calcutta   
1          http://www.flickr.com/photos/britishlibrary/ta...    London   
2          http://www.flickr.com/photos/britishlibrary/ta...    London   
3          http://www.flickr.com/photos/britishlibrary/ta...    London   
4          http://www.flickr.com/photos/britishlibrary/ta...    London   

      issuance                                            authors  date  \
0  monographic                             {'creator': ['A. A.']}  1841   
1  monographic                    {'creator': ['Abbott, Evelyn']}  1888   
2  monographic  {'contributor': ['LEECH, John - Artist'], 'cre...  1847   
3  monographic           {'creator': ['Laffan, De Courcy - Mrs']}  1892   
4  monographic  {'creator': ['Adams, W. H. Davenport (William ...  1863   

                       pdf identifier corporate  \
0  {'1': 'lsidyv35c55757'}  000000196        {}   
1  {'1': 'lsidyv376da437'}  000004047        {}   
2  {'1': 'lsidyv38b27a31'}  000005382        {}   
3  {'1': 'lsidyv3c4a946a'}  000014627        {}   
4  {'1': 'lsidyv32c34c22'}  000017057        {}   

                          fulltext_filename  \
0  sample/full_texts/000000196_01_text.json   
1  sample/full_texts/000004047_01_text.json   
2  sample/full_texts/000005382_01_text.json   
3  sample/full_texts/000014627_01_text.json   
4  sample/full_texts/000017057_01_text.json   

                                                imgs  
0                                                NaN  
1                 {'0': {'000257': ['11104648374']}}  
2  {'0': {'000410': ['11026944866'], '000412': ['...  
3  {'0': {'000134': ['11300364604'], '000039': ['...  
4                 {'0': {'000005': ['11046098454']}}

Since reading from files is a very common operation in any data analysis workflow, `pandas` provides methods to read from a variety of formats (JSON, CSV, clipboard, etc.)

The block of code above can be replaced by the following one-liner:

In [213]:
books_df = pd.read_json(json_file_path)

/var/folders/h0/0v0s48pj41zdnzdy7333vb4m0000gn/T/ipykernel_55035/3298360895.py:1: FutureWarning: The behavior of 'to_datetime' with 'unit' when parsing strings is deprecated. In a future version, strings will be parsed as datetime strings, matching the behavior without a 'unit'. To retain the old behavior, explicitly cast ints or floats to numeric type before calling to_datetime.
  books_df = pd.read_json(json_file_path)


In [214]:
books_df.head()

datefield                                         shelfmarks  \
0                1841               [British Library HMNTS 11601.ddd.2.]   
1                1888                [British Library HMNTS 9025.cc.14.]   
2  1847, 48 [1846-48]  [British Library HMNTS C.131.d.16., British Li...   
3                1892               [British Library HMNTS 10351.cc.60.]   
4                1863                 [British Library HMNTS 9078.c.10.]   

               publisher                                              title  \
0      Privately printed  [The Poetical Aviary, with a bird's-eye view o...   
1             Rivingtons  [A History of Greece. Part I. From the earlies...   
2           Punch Office  [The Comic History of England ... With ... col...   
3  Eden, Remington & Co.  [The Cruise of “The Tomahawk”: the story of a ...   
4     Virtue Bros. & Co.        [Scenes from the Drama of European History]   

  edition                          flickr_url_to_book_images     place  \
0          http://www.flickr.com/photos/britishlibrary/ta...  Calcutta   
1          http://www.flickr.com/photos/britishlibrary/ta...    London   
2          http://www.flickr.com/photos/britishlibrary/ta...    London   
3          http://www.flickr.com/photos/britishlibrary/ta...    London   
4          http://www.flickr.com/photos/britishlibrary/ta...    London   

      issuance                                            authors  \
0  monographic                             {'creator': ['A. A.']}   
1  monographic                    {'creator': ['Abbott, Evelyn']}   
2  monographic  {'contributor': ['LEECH, John - Artist'], 'cre...   
3  monographic           {'creator': ['Laffan, De Courcy - Mrs']}   
4  monographic  {'creator': ['Adams, W. H. Davenport (William ...   

                 date                      pdf  identifier corporate  \
0 1970-01-01 00:30:41  {'1': 'lsidyv35c55757'}         196        {}   
1 1970-01-01 00:31:28  {'1': 'lsidyv376da437'}        4047        {}   
2 1970-01-01 00:30:47  {'1': 'lsidyv38b27a31'}        5382        {}   
3 1970-01-01 00:31:32  {'1': 'lsidyv3c4a946a'}       14627        {}   
4 1970-01-01 00:31:03  {'1': 'lsidyv32c34c22'}       17057        {}   

                          fulltext_filename  \
0  sample/full_texts/000000196_01_text.json   
1  sample/full_texts/000004047_01_text.json   
2  sample/full_texts/000005382_01_text.json   
3  sample/full_texts/000014627_01_text.json   
4  sample/full_texts/000017057_01_text.json   

                                                imgs  
0                                                NaN  
1                 {'0': {'000257': ['11104648374']}}  
2  {'0': {'000410': ['11026944866'], '000412': ['...  
3  {'0': {'000134': ['11300364604'], '000039': ['...  
4                 {'0': {'000005': ['11046098454']}}

#### From CSV

Similarly to `pandas.read_json()`, `pandas.read_csv()` is there to make your life easier when it comes to loading CSV data into a dataframe (and that happens very often!).

This is particularly useful if you want to export dataframes into files compatible with Excel or other tabular data software. 

Let's see how to import one of the CSV files from the "Venice Apprenticeship" dataset (`../data/apprenticeship_venice/`). 

They contain information extracted from ~10,000 work contracts in 17th-century Venice, in particular master-apprentice relationships in the glass-industry. The apprentices were nick-named *garzoni*, hence the name of the dataset.

In [215]:
csv_file_path = '../data/apprenticeship_venice/professions_data.csv'

Try the following. Does it work? Before going to the next cell, can you guess why?

In [216]:
garzoni_df = pd.read_csv(csv_file_path)

ParserError: Error tokenizing data. C error: Expected 7 fields in line 36, saw 8


Let's have a look at the file first:

In [217]:
!head -n 2 ../data/apprenticeship_venice/professions_data.csv

page_title;register;annual_salary;a_profession;profession_code_strict;profession_code_gen;profession_cat;corporation;keep_profession_a;complete_profession_a;enrolmentY;enrolmentM;startY;startM;length;has_fled;m_profession;m_profession_code_strict;m_profession_code_gen;m_profession_cat;m_corporation;keep_profession_m;complete_profession_m;m_gender;m_name;m_surname;m_patronimic;m_atelier;m_coords;a_name;a_age;a_gender;a_geo_origins;a_geo_origins_std;a_coords;a_quondam;accommodation_master;personal_care_master;clothes_master;generic_expenses_master;salary_in_kind_master;pledge_goods_master;pledge_money_master;salary_master;female_guarantor;period_cat;incremental_salary
Carlo Della sosta (Orese) 1592-08-03;asv, giustizia vecchia, accordi dei garzoni, 114, 155;NA;orese;orese;orefice;orefice;Oresi;1;1;1592;08;1592;08;3;0;orese;orese;orefice;orefice;Oresi;1;1;1;Zuan Battista;Amigoni;;;0, 0;Carlo Della sosta;17;1;;;0, 0;1;0;1;1;1;0;0;0;0;0;NA;0


More than a comma-separated value, it looks like semicolon-separated values...
We thus need to adjust the `sep` parameter to specify which character is used to separate column values.

In [218]:
garzoni_df = pd.read_csv(
    csv_file_path,
    sep=';'
)

### Export

Once you have a `DataFrame` to play with in you python environment, exporting it is quite straightforward. 

Each export format has its own method: `.to_csv()`, `.to_json()`, `.to_html()`, etc.

The argument of the function is simply the `<path name>`, and sometime optional arguments, like which value separator you want to use for `.csv` files:

`<your_dataframe>.to_csv(<export_path>, sep=<which_separator>)`

In [221]:
toy_df.to_csv('example_dataframe.csv', sep=';')

## Section (4): Accessing and manipulating data.

We now have learned to get our hands on larger dataframes, with thousands of rows and dozens of columns, akin to the sort you may manipulate in a typical DH project.

It is now time to learn how to access the data stored in those dataframes.

### Indexing

There are two main ways to find information contained in a dataframe cell:
- either you know exactly where it is in the dataframe, for example if your frame is in a specific order;
- or, you want to access one or more cells based on conditions.

Let's take our books dataframe for the first of those cases. If you want to know the 'title' of the n-th row, you can get it using the `.at` keyword. The structure is:

`<dataframe>.at[<row_number>, <column_name>]`

✏️ [Ex.9]
- ✏️ What is the 'title' of the 26th row?


If we take the example of the `books_df` again, we might want to find all books that have been printed in 1841, or perhaps do we want all books printed privately, or perhaps still do we want all books privately printed in 1841:

The property to use to do this is called `.loc`. This is perhaps the most important `pandas` function.

It is applied to a `pandas.DataFrame` and returns a subset of that dataframe that matches the conditions given.

What is the structure? Retaking the examples phrased earlier, this is what it looks like? Is that structure clear to you?

In [242]:
# One condition
messy_data = books_df.loc[~(books_df['datefield'].str.isnumeric())]
clean_data = books_df.loc[(books_df['datefield'].str.isnumeric())]

print('Clean data: ', len(clean_data))
print('Messy data: ', len(messy_data))
print(f"The data to clean is: {100*len(messy_data)/len(books_df)}%")

Clean data:  370
Messy data:  82
The data to clean is: 18.141592920353983%


In [256]:
messy_data

datefield                                         shelfmarks  \
2    1847, 48 [1846-48]  [British Library HMNTS C.131.d.16., British Li...   
13          1893 [1892]               [British Library HMNTS 012641.i.62.]   
19              [1885.]               [British Library HMNTS 11781.df.24.]   
25              [1860.]            [British Library HMNTS 11781.d.23.(3.)]   
27               [1837]  [British Library HMNTS 1061.k.10., British Lib...   
..                  ...                                                ...   
435              [1894]  [British Library HMNTS 9061.ccc.23., British L...   
439             [1898.]           [British Library HMNTS 11603.ee.29.(4.)]   
443              1820?]  [British Library HMNTS 11642.de.33., British L...   
445           1881, etc                 [British Library HMNTS 9502.c.21.]   
449              [1833]          [British Library HMNTS 10347.ee.18.(14.)]   

                           publisher  \
2                       Punch Office   
13                          E. Stock   
19           London Literary Society   
25   Published & sold in the Theatre   
27                     Durrie & Peck   
..                               ...   
435       The African Review Offices   
439   “Home Words” Publishing Office   
443                                    
445                       Gill & Son   
449                         Atkinson   

                                                 title edition  \
2    [The Comic History of England ... With ... col...           
13                            [As the Angels. A story]           
19   [Thespis and Melpomene. A lyrical drama, in fo...           
25   [Lurline: a grand romantic original opera, in ...           
27   [Connecticut Historical Collections, containin...           
..                                                 ...     ...   
435  [The Downfall of Lobengula: the cause, history...           
439                            [“Resurgam” and Lyrics]           
443                 [Poems. [MS. note by the author.]]           
445    [The Intermediate Education History of England]           
449  [Address to the Friends of Justice and Humanit...           

                             flickr_url_to_book_images              place  \
2    http://www.flickr.com/photos/britishlibrary/ta...             London   
13   http://www.flickr.com/photos/britishlibrary/ta...             London   
19   http://www.flickr.com/photos/britishlibrary/ta...             London   
25   http://www.flickr.com/photos/britishlibrary/ta...             London   
27   http://www.flickr.com/photos/britishlibrary/ta...          New Haven   
..                                                 ...                ...   
435  http://www.flickr.com/photos/britishlibrary/ta...             London   
439  http://www.flickr.com/photos/britishlibrary/ta...             London   
443  http://www.flickr.com/photos/britishlibrary/ta...  Privately printed   
445  http://www.flickr.com/photos/britishlibrary/ta...             Dublin   
449  http://www.flickr.com/photos/britishlibrary/ta...           Bradford   

        issuance                                            authors  \
2    monographic  {'contributor': ['LEECH, John - Artist'], 'cre...   
13   monographic                                                 {}   
19   monographic                      {'creator': ['ARTERTON, A.']}   
25   monographic  {'contributor': ['WALLACE, William Vincent - M...   
27   monographic              {'creator': ['BARBER, John Warner.']}   
..           ...                                                ...   
435  monographic  {'contributor': ['COLLINGRIDGE, Leonard Thomas...   
439  monographic             {'creator': ['WINSCOMBE, John Cave.']}   
443  monographic                {'creator': ['WRANGHAM, Francis.']}   
445  monographic                     {'creator': ['WREN, Edmond.']}   
449  monographic                                                 {}   

                   date  

In [243]:
# Another condition
books_df.loc[books_df['publisher']=='Privately printed']
books_df.loc[books_df['datefield']=='1841']


datefield                                 shelfmarks          publisher  \
0        1841       [British Library HMNTS 11601.ddd.2.]  Privately printed   
284      1899  [British Library HMNTS 11601.ddd.11.(6.)]  Privately printed   

                                                 title edition  \
0    [The Poetical Aviary, with a bird's-eye view o...           
284                            [Nightshades. [Poems.]]           

                             flickr_url_to_book_images     place     issuance  \
0    http://www.flickr.com/photos/britishlibrary/ta...  Calcutta  monographic   
284  http://www.flickr.com/photos/britishlibrary/ta...     Paris  monographic   

                                               authors                date  \
0                               {'creator': ['A. A.']} 1970-01-01 00:30:41   
284  {'creator': ['MONKSHOOD, G. F. - pseud. [i.e. ... 1970-01-01 00:31:39   

                         pdf  identifier corporate  \
0    {'1': 'lsidyv35c55757'}         196        {}   
284  {'1': 'lsidyv37212e0a'}     2527441        {}   

                            fulltext_filename imgs  
0    sample/full_texts/000000196_01_text.json  NaN  
284  sample/full_texts/002527441_01_text.json  NaN

In [247]:
# Both conditions
books_df.loc[(books_df['datefield']=='1841')&(books_df['publisher']=='Privately printed')]

datefield                            shelfmarks          publisher  \
0      1841  [British Library HMNTS 11601.ddd.2.]  Privately printed   

                                               title edition  \
0  [The Poetical Aviary, with a bird's-eye view o...           

                           flickr_url_to_book_images     place     issuance  \
0  http://www.flickr.com/photos/britishlibrary/ta...  Calcutta  monographic   

                  authors                date                      pdf  \
0  {'creator': ['A. A.']} 1970-01-01 00:30:41  {'1': 'lsidyv35c55757'}   

   identifier corporate                         fulltext_filename imgs  
0         196        {}  sample/full_texts/000000196_01_text.json  NaN

Conditions can be combined using Boolean logic. We can thus for example search for:
- Condition A **and** Condition B
- Condition A **or** Condition B
- Condition A **and not** Condition C
- etc.

The boolean operators are the same as in traditional Python: `&`, `|`, `~`, etc.

✏️ [Ex.10]

Using the indexing structure we've just seen, try to find which book(s) obey the following conditions:

- ✏️ 1° Published by 'John Murray' in 1856
- ✏️ 2° Printed in London in 1818
- ✏️ 3° Published between 1830 and 1848
- ✏️ 4° Published by 'Longmans & Co.' or 'Henry Colburn'
- ✏️ 5° Published in 1894 but not by 'W. Blackwood & Sons'

In [261]:
# your solution
subset_books = books_df.loc[books_df['datefield'].str.isnumeric()]
subset_books['datefield'] = subset_books['datefield'].astype(int)

/var/folders/h0/0v0s48pj41zdnzdy7333vb4m0000gn/T/ipykernel_55035/3698726572.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  subset_books['datefield'] = subset_books['datefield'].astype(int)


### NaN values

Throughout your own research, you will most likely stumble upon instances where your data is incomplete: incomplete metadata from an external source, data manipulation exceptions that don't return a value, missing record. 

This will take the form of a cell containing a `None` or `numpy.nan` value.

It is important to know how to filter them out (or in) in `pandas`, especially as they can mess up some functions you might want to apply to the whole dataframe.

For example, let's take our earlier toy dataframe and remove one cell;

In [265]:
toy_df

date        event       day  is_weekend  is_weekday
0 1970-05-23        birth  Saturday        True       False
1 1978-07-14  anniversary    Friday       False        True
2 1986-03-14      wedding    Friday       False        True
3 1993-01-01      wedding    Friday       False        True
4 1998-07-14  anniversary   Tuesday       False        True

In [266]:
toy_df.at[2, 'event'] = np.nan
toy_df

date        event       day  is_weekend  is_weekday
0 1970-05-23        birth  Saturday        True       False
1 1978-07-14  anniversary    Friday       False        True
2 1986-03-14          NaN    Friday       False        True
3 1993-01-01      wedding    Friday       False        True
4 1998-07-14  anniversary   Tuesday       False        True

To find which rows are NA in a specific **column**, we use the following `.isna()` method.

It returns a `series` of booleans telling whether the matching row has a nan value.

In [267]:
toy_df['event'].isna()

0    False
1    False
2     True
3    False
4    False
Name: event, dtype: bool

Using our `.loc` operator, we can convert that to a subset of the dataframe:

In [268]:
toy_df.loc[toy_df['event'].isna()]

date event     day  is_weekend  is_weekday
2 1986-03-14   NaN  Friday       False        True

Empty cells can be filled with one replacement value. For example here, we might want to prefer "unknown" to a NaN value.

Be careful, `.fillna()` returns a `pandas.Series`, not a full dataframe. Therefore we use its output to recreate a column:

In [277]:
s = toy_df['event'].fillna('not known', inplace=True)

/var/folders/h0/0v0s48pj41zdnzdy7333vb4m0000gn/T/ipykernel_55035/2259728707.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  type(toy_df['event'].fillna('not known', inplace=True))


NoneType

### Reindexing

To change the value of one or more cells, you should know which rows are concerned by the change.
Like earlier, this can come either:
- from the precise row number (if you know beforehand which row/column cell you want to change)
- from a certain condition being met (let's change all cells that start with 'unk', for example)

The reindexing can be thought of as two distinct parts:
- 1° knowing which rows to change;
- 2° modifying the values of these rows for a specific column.

To do this, use the following structure:

`<dataframe>.loc[<dataframe>[<column_to_test>]==<value_to_check_for>, <column_to_change>] = <new_value>`

Continuing with our toy dataframe, this expression selects all rows whose 'event' is 'birth':

In [282]:
toy_df.loc[toy_df['event']=='wedding']

date    event     day  is_weekend  is_weekday
3 1993-01-01  wedding  Friday       False        True

Let's say you may want to change that into 'marriage':

In [283]:
toy_df.loc[toy_df['event']=='wedding', 'event'] = 'marriage'

In [284]:
toy_df

date        event       day  is_weekend  is_weekday
0 1970-05-23        birth  Saturday        True       False
1 1978-07-14  anniversary    Friday       False        True
2 1986-03-14    not known    Friday       False        True
3 1993-01-01     marriage    Friday       False        True
4 1998-07-14  anniversary   Tuesday       False        True

But again, the testing column doesn't have to be the one you change.

More generally, this is:
`<dataframe>.loc[<condition>, <column_to_change>] = <new_value>`

For example, say I want to change my last name in our earlier dataframe:

✏️ [Ex. 11]

In this exercise, we will combine what we've seen on Indexing, Detecting NA values, and Reindexing.

The dataframe of books that was opened from a JSON earlier `books_df` contains information on which is the identifier to a first page scan. This information is in the column `imgs`.
In some cases however, the image does not exist and the cell is empty.

- ✏️ Find how many books do not have an image (column `imgs`)
- ✏️ Find *which* books do not have an image
- ✏️ Replace the empty cells that do not have an image with a dummy value
- ✏️ (Advanced) Find which are the top four geographical origin for books without an image.

In [312]:
books_df.loc[books_df['imgs'].isna(), 'imgs'] = 'sven.com'

### Iterating

As a rule of thumb, one should prioritise any data manipulation that can be performed in one shot. `pandas.DataFrame` are ineffecient objects for frequent modification. 

However, there are scenarios where you may want to iterate over the rows of the dataframe. For example, to perform row-wise tests or manipulation that require access to more than one column.

To do that, the `.iterrows()` method (for "iterration over rows") should be used. 

Just like an `enumerate` in traditional python, the `.iterrows()` method will return two elements at each iteration:
- the index of the row (between 0 and the length of the dataframe);
- the row corresponding to that iteration.

That row is a `pandas.Series`. It acts a bit like a python `dictionary` and the value for a specific column is accessed through:
`row[<column>]`



In [311]:
for i in toy_df.iterrows():
    index = i[0]
    row = i[1]
    


In [313]:
for index, row in toy_df.iterrows():
    print('The event at index', index, 'is:', row['event'])

The event at index 0 is: birth
The event at index 1 is: anniversary
The event at index 2 is: not known
The event at index 3 is: marriage
The event at index 4 is: anniversary


✏️ [Ex. 12]

Here we will combine iteration, and `datetime` and `str` accessors. By iterating over the `toy_df` dataframe:

- ✏️ Print the date of each row as well
- ✏️ Print the day of the week that each date corresponds to
- ✏️ Print whether the event starts with 'bir'

In [316]:
# your solution

for i in toy_df.iterrows():
    index = i[0]
    row = i[1]

    print(row['date'])
    print(row['day'])
    print(row['event'].startswith('bir'))
    print('----')

1970-05-23 00:00:00
Saturday
True
----
1978-07-14 00:00:00
Friday
False
----
1986-03-14 00:00:00
Friday
False
----
1993-01-01 00:00:00
Friday
False
----
1998-07-14 00:00:00
Tuesday
False
----


## Final exercise: Shakespeare and Company project

**Dataset**

For this excercise, we will be working with an open-access data from a DH research project: the [*Shakespeare and Company project*](https://shakespeareandco.princeton.edu/).

The dataset we'll be using contains the list of books that were lent out in the 20th century by the celebrated *Shakespeare and Company* library in Paris.

The dataset can be downloaded from the following address:https://dataspace.princeton.edu/handle/88435/dsp01jm214s28p (file size = ~2MB).

You may use CSV or JSON:
- the CSV file will only contain `str`, `float` and `int`;
- whereas the JSON file will contain `lists`.

The choice will be one of comfort. Pick whichever yields the datatypes you are more comfortable with.
 
**Steps**

Start by loading the file it into a `pandas.DataFrame` called "SCo"

**Try to answer the following questions**
- 1° How many records does it contain?


- 2° What's the format(s) of the **most purchased** document(s)? How many times was it/were they purchased?
- 3° What's the format(s) of the **least borrowed** document(s)? How many times was it/were they borrowed?

- 4° Which is the most borrowed **book**? How many times was it borrowed?

- 5° How many 'Poems' are there in collection? What is the earliest one? How many don't have a date?

- 6° Replace the empty cells in the `circulation_year` column with "unknown".
- 7° Find which books were in circulation in 1951
- 8° (Advanced) By iterating over the rows of the dataframe, create a dictionnary of dates that gives how many books from the datafram were in circulation that year.

In [317]:
sco = pd.read_csv('../data/SCoData_books_v1.2_2022_01.csv')

In [359]:
sco_json = pd.read_json('../data/SCoData_books_v1.2_2022_01.json')

In [337]:
#Question 4: 

# Step 1 Finding the maximum number of purchases
max_count = sco.loc[(sco['format']=='Book'), 'borrow_count'].max()

# Step 2 Findinding which books have been purchased
most_purchased = sco.loc[(sco['borrow_count'] == max_count)&(sco['format']=='Book')]

#Step 3 Which formats?
most_purchased['format'].value_counts()

56

In [353]:
#Question 5

#Step 1: Find all books called 'Poems'

poems = sco.loc[(sco['title']=='Poems')]
#Step 2: What is earliest?
poems.sort_values('year', ascending=True).head(1)

#Step3: How many don't have a date?
len(poems.loc[poems['year'].isna()])


16

In [354]:
# Question 6

sco['circulation_years'] = sco['circulation_years'].fillna('unknown')

In [363]:
# Question 7

#1. CSV
sco.loc[sco['circulation_years'].str.contains('1951')]

#2. JSON
sco_json.loc[sco_json['circulation_years'].apply(lambda x: 1951 in x)]

uri  \
1     https://shakespeareandco.princeton.edu/books/m...   
1121  https://shakespeareandco.princeton.edu/books/l...   
1271  https://shakespeareandco.princeton.edu/books/r...   
3953  https://shakespeareandco.princeton.edu/books/s...   
4096  https://shakespeareandco.princeton.edu/books/r...   

                                 title                    author    year  \
1                               Utopia            [More, Thomas]  1516.0   
1121      The Pageant of English Prose                       NaN  1912.0   
1271      Pointed Roofs (Pilgrimage 1)  [Richardson, Dorothy M.]  1915.0   
3953              All Men Are Brothers             [Shi, Nai'an]  1933.0   
4096  The English Vision: An Anthology                       NaN  1933.0   

     format  uncertain                                          ebook_url  \
1      Book      False         https://archive.org/details/utopiamo00more   
1121   Book      False  https://archive.org/details/thepageantof00leon...   
1271   Book      False  https://archive.org/details/pointedroofs00rich...   
3953   Book      False                                                NaN   
4096   Book      False  https://archive.org/details/in.ernet.dli.2015....   

      event_count  borrow_count  purchase_count  \
1               2             2               0   
1121            2             2               0   
1271           50            45               2   
3953            5             5               0   
4096            2             2               0   

                                      circulation_years  \
1                                    [1930, 1950, 1951]   
1121                                       [1938, 1951]   
1271  [1921, 1922, 1923, 1926, 1927, 1931, 1932, 193...   
3953                           [1936, 1938, 1939, 1951]   
4096                                       [1940, 1951]   

                               updated volumes_issues notes            editor  \
1     2021-10-31T17:47:49.103296+00:00            NaN   NaN               NaN   
1121         2021-01-27T21:33:31+00:00            NaN   NaN  [Leonard, R. M.]   
1271         2021-04-17T21:48:48+00:00            NaN   NaN               NaN   
3953         2020-07-17T00:55:56+00:00            NaN   NaN               NaN   
4096         2021-06-05T01:11:24+00:00            NaN   NaN   [Read, Herbert]   

            translator introduction photographer illustrator  
1                  NaN          NaN          NaN         NaN  
1121               NaN          NaN          NaN         NaN  
1271               NaN          NaN          NaN         NaN  
3953  [Buck, Pearl S.]          NaN          NaN         NaN  
4096               NaN          NaN          NaN         NaN

In [374]:
#Question 8

year_dict = {}

for i in sco.iterrows():
    index = i[0]
    row = i[1]

    year_str = row['circulation_years']
    

    list_of_years = [int(i) for i in year_str.split(';') if i!='unknown']

    for year in list_of_years:
        if year in year_dict.keys():
            year_dict[year] +=1
        else:
            year_dict[year] =1
        
        